woocommerce/woocommerce-ios
# Add-Ons: Parse add-ons from product metadata

part of #3886

# Why

As part of showing requested add-ons for an order, we first need to parse the add-ons from the product response.

This PR takes care of converting that add-on info into a `ProductAddOn` networking entity. Which also adds some new entities as dependencies:

- ProductAddOn
- ProductAddOnOption
- AddOnType
- AddOnTitleFormat
- AddOnRestrictionsType
- AddOnPriceType
- AddOnDisplay

# How

- The add-on parsing is fairly straightforward, we are just declaring the necessary `decodable` structs to let swift do its thing.

- The complicated bit is on accessing the add-on JSON objects from the product `metadata`, as the `metadata` is not a typed object and it is wrapped in not ideal JSON structure(Array with objects, where the "key" key is "_product_addons" and its "value" key has the array of add-ons).

To isolate that complexity I've created a `ProductAddOnEnvelope` type that provides a `resolve()` function to try to extract those add-ons.

```swift
"meta_data": [
 {
  "id": 6469,
  "key": "_last_editor_used_jetpack",
  "value": "classic-editor"
},
{
  "id": 6471,
  "key": "_product_addons",
  "value": [
    {
      "description": "Pizza topping",
      "title_format": "label",
      "required": 0,
    }
  ]
}
]
```

- Ran `rake generate` to generate `fakes` and `copiable` methods for the new types

# Testing Steps
- I've added a new test case to check that the above JSON parses correctly, so making sure the test pass should be enough for now! 
 

Update release notes:
